src/Microsoft.OData.Edm/Schema/EdmTargetPath.cs

```diff
@@ -0,0 +1,111 @@
+//---------------------------------------------------------------------
+// <copyright file="EdmTargetPath.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.OData.Edm
+{
+    /// <inheritdoc/>
+    public class EdmTargetPath : IEdmTargetPath
+    {
+        private List<IEdmElement> segments;
+        private readonly string path;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EdmTargetPath"/> class.
+        /// </summary>
+        /// <param name="segments">Target path segments.</param>
+        public EdmTargetPath(IEnumerable<IEdmElement> segments)
+        {
+            EdmUtil.CheckArgumentNull(segments, nameof(segments));
+
+            this.segments = segments.ToList();
+
+            ValidateSegments();
+
+            this.path = this.GetPathString();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EdmTargetPath"/> class.
+        /// </summary>
+        /// <param name="segments">Target path segments.</param>
+        public EdmTargetPath(params IEdmElement[] segments)
+            :this(segments.AsEnumerable())
+        {
+        }
+

```

Comment:
I'd properse adding an `internal EdmTargetPath` construcor that accepts a path string for those scenarios where we construct an `EdmTargetPath` while we already have the path as input. If necessary, we could add some validation behing a `#if DEBUG` flag to verify the segments match those in the path string.

Reply:
I think Clement's suggestion to add internal constructor only takes `string path` as parameter, right?

Reply:
An internal constructor that takes both `IEnumerable<IEdmElement>` and `string path`. This is because when we already have the `string path`, then compute the `IEnumerable<IEdmElement>`, we can initialize an `EdmTargetPath` without recomputing the `string path` again